### ISSUE TITLE
Review and improve maps

### ISSUE THREAD

**Issue by axelboc** [axelboc/anki-ultimate-geography#1]
- Check quality and file size of all maps and replace the worst ones.
- Prefer SVG or PNG format rather than JPEG.
- Add source URLs and attributions to LICENCE file.

**Comment by axelboc** [comment]
Experiment with hashes in SVG URLs:

1. Select a few regional SVG maps from Wikipedia
2. Define a path for each country and give it an ID
3. Add `:target { fill: <colour>; }` to the CSS of the Anki cards or of the SVGs
4. Set the `src` attribute of the image in the map field of each note to `<regional-map>.svg#<country-id>`

This could make the deck a lot lighter, as regional maps would not have to be duplicated. It would definitely guarantee consistency across all maps. It may not, however, be supported by the web view that Anki uses.

**Comment by axelboc** [comment]
I replaced most of the old, inconsistent JPEG maps with PNGs: 2be14042088cc874de94604353a47952dbced98e

The new maps are from the German [Locator map](https://de.wikipedia.org/wiki/Wikipedia:Kartenwerkstatt/Positionskarten) project of Wikipedia's [Map workshop](https://en.wikipedia.org/wiki/Wikipedia:Graphics_Lab/Map_workshop). More specifically, they are 500px-wide PNG versions of SVG locator maps from the project's [16:9 regional location map scheme](https://commons.wikimedia.org/wiki/Category:SVG_locator_maps_(16:9_regional_location_map_scheme)). The SVG maps were too big to be included directly, so I downloaded the PNG versions instead and further optimised them with [PNGGauntlet](https://pnggauntlet.com/). Here are a couple of the new maps:

![ug-map-bulgaria](https://cloud.githubusercontent.com/assets/2936402/22314929/de66c45e-e3b7-11e6-859c-1ca5a9d25efc.png)
![ug-map-british_virgin_islands](https://cloud.githubusercontent.com/assets/2936402/22314930/de69d842-e3b7-11e6-9ebc-a11a95378587.png)

Unfortunately, the locator map project is not complete and does not cover some of the maps included in this deck. The maps that **still need improvement** are:

1. The maps of **oceans, seas, straits, bays and gulfs**. I have my eyes on a category called [Locator maps of water bodies](https://commons.wikimedia.org/wiki/Category:Locator_maps_of_water_bodies) which is marked for clean up and may provide a nice, consistent and more exhaustive set of maps in the future.
2. The maps of **Azores, Ceuta, Corsica, Galapagos Islands, Kaliningrad Oblast, Melanesia, Melilla, Micronesia, Polynesia, Saint Helena, Sardinia, Sumatra and Zanzibar**, which were either absent from the 16:9 set, not in scope of the locator map project, or not good or precise enough to justify their inclusion when compared to the current maps.
3. Many locator maps of **islands and archipelagos** taken from the 16:9 set, which could use a zoom box for added clarity (and to make them more distinguishable from islands in the same region). It would help a lot as well if the borders between territories were represented. Some of the Pacific islands maps could also be a bit more zoomed out to give them more context. Here are a couple of maps that could be improved, followed by one that is satisfactory:

![ug-map-united_states_virgin_islands](https://cloud.githubusercontent.com/assets/2936402/22316049/9ccb941e-e3be-11e6-804d-693d1ca8f85a.png)
![ug-map-pitcairn_islands](https://cloud.githubusercontent.com/assets/2936402/22316052/aab0db52-e3be-11e6-8c18-fb374649c373.png)
![ug-map-antigua_and_barbuda](https://cloud.githubusercontent.com/assets/2936402/22316077/c9011806-e3be-11e6-92c5-506b4143e756.png)

**Comment by splintax** [comment]
I was finding it impossible to learn the Lesser Antilles with the UG map set. [Wikipedia:Blank maps](https://en.wikipedia.org/wiki/Wikipedia:Blank_maps) has some blank SVG maps at different scales we could use instead. I've manually edited a few cards in my personal deck, eg. these two maps to distinguish the Virgin Islands:

# UG default (30kb PNG)

![ug-map-british_virgin_islands](https://user-images.githubusercontent.com/1313976/45824304-07676b00-bd33-11e8-85fe-3a5bd8b4c50b.png)
![ug-map-united_states_virgin_islands](https://user-images.githubusercontent.com/1313976/45824314-10583c80-bd33-11e8-8f0f-f86787c21f89.png)

# Modified Wikipedia SVGs (107kb SVG, rendered as 300 dpi PNG because GitHub doesn't support SVG in comments)

![antilles - us virgin islands](https://user-images.githubusercontent.com/1313976/45824395-3ed61780-bd33-11e8-9ec9-299a38cc602f.png)
![antilles - british virgin islands](https://user-images.githubusercontent.com/1313976/45824509-83fa4980-bd33-11e8-9bc4-f7585c808a9e.png)

It would be nice if we could replace all the maps with clean SVGs with consistent colours, and a zoom level that responds to the size of the country. I'm not too sure how to do it programmatically though.

**Comment by splintax** [comment]
I've created a version of that SVG where each country is on a separate named layer (ie. wrapped in `<g id="Country_Name">`). This code works in Firefox to highlight it programmatically:

    <object id="antilles" data="antilles.svg"></object>
    <script type="text/javascript">
        var highlightCountry = function(country, color) {
            var svg = document.getElementById("antilles").contentDocument
            for (var node of svg.getElementById(country).children) {
                if ("style" in node) {node.style.fill = color}
            }
        }
        window.onload = function() {highlightCountry("Antigua_and_Barbuda", "blue")}
    </script>

Unfortunately, I couldn't figure out a pure CSS solution and the JavaScript doesn't work in Anki – something to do with attempting to access the `contentDocument` property on an embedded SVG referenced from a different file. Another method might be to inject a stylesheet into the SVG document rather than manually iterating over its children. I was struggling to figure this out without a debugger, but I've just realised Anki has a debug console, so I'll report back if I make any progress. In the meantime, if anyone knows how Anki's JavaScript engine works please let me know.

**Comment by splintax** [comment]
Problem solved! I have replaced the following files:

* ug-map-anguilla.png
* ug-map-antigua_and_barbuda.png
* ug-map-aruba.png
* ug-map-barbados.png
* ug-map-bonaire.png
* ug-map-british_virgin_islands.png
* ug-map-curacao.png
* ug-map-dominica.png
* ug-map-grenada.png
* ug-map-guadeloupe.png
* ug-map-martinique.png
* ug-map-montserrat.png
* ug-map-puerto_rico.png
* ug-map-saba.png
* ug-map-saint_kitts_and_nevis.png
* ug-map-saint_lucia.png
* ug-map-saint_vincent_and_the_grenadines.png
* ug-map-sint_eustatius.png
* ug-map-sint_maarten.png
* ug-map-trinidad_and_tobago.png
* ug-map-united_states_virgin_islands.png

... with [Lesser_Antilles_location_map.svg](view-source:https://upload.wikimedia.org/wikipedia/commons/2/22/Lesser_Antilles_location_map.svg) (I uploaded my version of this image, with semantic IDs and classes, to Wikipedia – note the CSS rule `.highlighted path, .highlighted polygon {fill: #f00;}`). Then, I added the following script to the end of the SVG file:

```javascript
window.addEventListener("message", function (msg) {
    document.getElementById(msg.data).classList = "highlighted"
}, false);
```

I then set the relevant field on the Anki note to the following, specifying an appropriate `COUNTRY_NAME`:

```html
<iframe id="antilles" src="antilles.svg"></iframe>
<script type="text/javascript">
var iframe = document.getElementById("antilles")
iframe.addEventListener("load", function (e) {
    iframe.contentWindow.postMessage("COUNTRY_NAME", "*")
})
</script>
```
And finally, I added the following CSS to the note template:

```css
#antilles {
  width: 100%;
  height: 100%;
  min-height: 50em;
  border: none;
}
```

I will continue to test this technique out but it seems to work well, and is a significant improvement over the standard maps. It should also be possible to use JavaScript to automatically generate enlarged overlays of the very small countries. By adding semantic IDs to 5 – 10 other maps on [Wikipedia:Blank_maps#SVG_format](https://en.wikipedia.org/wiki/Wikipedia:Blank_maps#SVG_format), it should be possible to replace all UG maps with a few vector images with consistent colours.

**Comment by axelboc** [comment]
Hey @splintax, sorry for the late reply. I've been keeping up with your work and I'm really glad you managed to improve the Caribbean maps in your deck!

I think the next step would be to make individual locator maps consistent with the ones already in the deck. The example below shows what I'm talking about ; it is a regional map of the Caribbean with a couple of additional features : a tiny world map in the bottom left corner for context, and a zoom box for precision.

![ug-map-antigua_and_barbuda](https://cloud.githubusercontent.com/assets/2936402/22316077/c9011806-e3be-11e6-92c5-506b4143e756.png)

I reckon your awesome _Lesser Antilles_ map could help generate those zoom boxes. I need to find the time to dig into the Locator map project's contributing guidelines to get this done once and for all. 😣

**Comment by splintax** [comment]
Hi @axelboc, I completely agree! I’ve actually been working on that here and there over the past week, and I’m almost done. I’m on holiday at the moment without access to my main computer, so I can’t share code or screenshots, but the `getBBox()` method makes most of the work easy: once you have put the country on its own layer, you can easily get the largest bounding box containing the entire country, clone it, and scale it up.

The tricky part is wrapping your head around how the SVG coordinate system works when you apply a scaling transform (the entire coordinate system is scaled, so generally your scaled drawing will move outside of the view box without an additional translation). Moreover, you need to figure out how much to scale and translate. I have taken this approach:

- get the coordinates of the four rectangles that are bounded by the edges of the map and colinear with one edge of the bounding box containing the country of interest,
- get the largest subrectangle of each rectangle which has the same aspect ratio as the bounding box, and select the largest such subrectangle (in the image above this would obviously be to the left of the selected country), and
- scale the cloned country to a fixed percentage (say 70%) of that subrectangle, to allow for some padding.

My code works for most of the countries on this map but breaks when the locator map needs to be to the right of the selected country. I hope to have time to finish it next week. What is the locator map project? It would be great if my code could be used outside my personal Anki deck.

There is another problem – the iframe solution described above doesn’t work in AnkiWeb or the Anki iOS app. I need to spend more time debugging it but it seems like iframes may not work at all – the image doesn’t even appear without executing the JavaScript. So I was more than a little overconfident when I wrote ‘problem solved’ earlier! For my own deck I’ll probably just use my code to generate JS-free SVGs, but this wastes storage space on AnkiWeb and bandwidth on the mobile network, so it’s not ideal. The SVG maps of Europe are over 1MB.

**Comment by axelboc** [comment]
Yeah, SVG maps are way too big. That's the reason why I've put 500px-wide, optimised PNG versions of the maps in the deck. They weigh at about 35 kB each on average, which is quite light, especially since we won't be adding but replacing maps. Thanks for helping with this, much appreciated! 🚀 Keep in touch 😉

**Comment by splintax** [comment]
I’ve started manually cleaning up [File:States_and_territories_of_Oceania.svg](https://commons.wikimedia.org/wiki/File:States_and_territories_of_Oceania.svg) so that it has country metadata and shares the same visual style as the Antilles map:

![locator-map-generator-oceania](https://user-images.githubusercontent.com/1313976/46911362-1a253680-cfa5-11e8-8c51-8a1c36ea8ebb.png)

It’s proving to be a more laborious than expected though, and even after trying to ‘crop’ and otherwise simplify the SVG, it's still around 200 KB – twice the size of the Antilles map. There are even bigger problems with [File:Blank map of Europe (with disputed regions).svg](https://en.wikipedia.org/wiki/File:Blank_map_of_Europe_(with_disputed_regions).svg), which is 799 KB (although it would probably need to be cropped to suit mobile Anki). Theoretically this work only needs to be done once, and all the locator map style and code is now in separate files, so the cleaned up SVGs can be usefully contributed back to Wikipedia. But I am worried about accidentally making cartographical errors as I manually add metadata, and it's not always possible to find the perfect SVG source data on Wikipedia anyway.

So, I think it might be worth automating the generation of the source SVGs as well. This would also give us global location and projection metadata which could be used to optionally add a scale, mark lines of latitude and longitude, or label geographical sub-regions, cities, and other landmarks. By judiciously cropping, simplifying un-zoomed paths, and minifying the result, it may be possible to generate SVGs that are no larger than the existing PNGs. I’m going to check out [mapshaper](https://mapshaper.org), [Polymaps](http://polymaps.org) and [Kartograph](https://kartograph.org).

**Comment by ohare93** [comment]
Any updates on this? I have now finished the deck, but the small Carribean and Oceania islands maps are killing me. I see a lot of fantastic progress by you lovely people in this thread, but are there finished map files to be imported?

**Comment by axelboc** [comment]
Hey, I hadn't made any progress at all on this, but your message got me motivated again. I've gone ahead and tried adding a zoom box to the map of Guadeloupe by hand with Inkscape. Here is what I did for future reference:

1. I started with the [Guadeloupe map](https://commons.wikimedia.org/wiki/File:Guadeloupe_in_its_region.svg).
1. I copied the zoom box from the [Falkland Islands map](https://commons.wikimedia.org/wiki/File:Falkland_Islands_in_its_region_(zoom).svg) then positioned it like on the [Antigua and Barbuda map](https://commons.wikimedia.org/wiki/File:Antigua_and_Barbuda_in_its_region_(zoomed).svg).
1. I adjusted the size of the red square around the islands of Guadeloupe to include a few more islands for context.
1. I copied the corresponding area from the [Lesser Antilles map](https://en.wikipedia.org/wiki/File:Lesser_Antilles_location_map.svg) and cropped the island and dividing line that didn't fit completely inside the zoom box.

Here is the result:

![ug-map-guadeloupe](https://user-images.githubusercontent.com/2936402/61410475-a376ea00-a8e4-11e9-9c20-06de88e0237a.png)

I'll add it to the deck asap. I'll do my best to improve some more maps as I find the time and motivation.

**Comment by aplaice** [comment]
To hopefully avoid duplicated efforts, I'm claiming the addition of zoom boxes to the remaining Lesser Antilles maps.

I've automated\* the selection of the "corresponding area" from the [Lesser Antilles Map](https://en.wikipedia.org/wiki/File:Lesser_Antilles_location_map.svg), for any arbitrary box in the X in its region (Caribbean map), which should speed things up.

If I don't finish or give a further update within a week, consider the task unclaimed. :)

\* no, I have no idea whether it was worth it...

**Comment by aplaice** [comment]
(And done, in https://github.com/axelboc/anki-ultimate-geography/pull/134.)

**Comment by aplaice** [comment]
Similarly to before: to avoid duplicated efforts (just in case somebody gets inspired at the same time), I'm claiming the overhauling of the sea maps. (For the moment I'm only claiming the ones "adjacent" to the Atlantic and Indian Oceans to avoid overreach, but since I'm using the almost magical [mapshaper](https://github.com/mbloch/mapshaper/) for automation, all of them should be  easily doable, modulo corner cases like some straits.)

Again, please consider the task unclaimed if I don't update within two weeks.

**Comment by axelboc** [comment]
How are you doing with this? Maybe don't go too far with the overhaul in case we decide, in #137, that some water bodies need to be removed from the deck (like straits). Anyway, very curious to hear more about how you're using this "almost magical mapshaper" tool!

**Comment by aplaice** [comment]
I've created the equivalent of the [country world map](https://commons.wikimedia.org/wiki/File:World_location_map_(W3).svg), but with the areas corresponding to all (almost all\*) the seas and some lakes also labelled with `id`s (not just the countries like in the original map), using data (shapefiles) from [Natural Earth](https://www.naturalearthdata.com/downloads/).

\* for some reason, Natural Earth doesn't seem to have the Celtic Sea in the main dataset. Otherwise everything seems to correspond to the [1953 IHO definitions](https://en.wikipedia.org/wiki/List_of_seas#cite_note-proposed-20). The seas are sometimes overly-split-up, so there are unnecessary borders, but no incorrect borders, which means that getting the correct sea highlighted will be just a matter of selecting several `id`s. (There's a nice shapefile [here](http://www.marineregions.org/downloads.php#iho), which appears to exactly (one-to-one) correspond to the IHO definitions, and which I've been using for comparison. Unfortunately it's under CC BY-NC-SA which is a no-go if the maps are to be ultimately also upstreamed to Wikimedia (which would be ideal). It's possible that since the shapefiles are essentially data, their license doesn't carry over to a map derived from them, but it might depend on jurisdiction and in any case I'm not keen to test the limits of copyright/copyleft law.)

The map (without any highlighting) looks like [this](https://gist.github.com/aplaice/652a47b70ff94d45924fa0039ed2daac) which is hopefully very similar to the style of the Wikimedia map that we have been using (or more precisely the Wikimedia map from which the country maps that we have been using were cropped).

The main part that remains is, for each sea, to highlight the correct areas and crop/zoom into the right part of the world.

For example, the Gulf of Mexico (WIP — I'll probably zoom in a bit more):

![gulf](https://user-images.githubusercontent.com/16461345/69008217-63824b00-0948-11ea-8749-b1586d8e64b8.png)

One "stretch goal" is to figure out how to "re-cut" the edges of the shapefiles — they're "pre-cut" at the anti-meridian — so that we could have maps centred at a longitude other than the prime-meridian. (`mapshaper` goes crazy — having lines crossing the map from edge to edge for any polygons at the edges — when I set a central longitude other than 0° for the Winkel III projection.) This would be particularly useful for the seas near the Pacific Ocean since the Winkel III centred at 0° deforms them considerably.

Mapshaper is "magical" as one can 

a) create an SVG map from the shapefiles with a single command, and

b) play around with/inspect the maps interactively

**Comment by Alexendoo** [comment]
I'm finding some of the Central Asian maps a bit tricky, e.g with Kyrgyzstan I have trouble telling that it borders China which adds a lot of context

![kyrgyzstan Map](https://raw.githubusercontent.com/axelboc/anki-ultimate-geography/fe897da1ab255cce52679716c17a3dc071e8c14c/src/media/ug-map-kyrgyzstan.png)

**Comment by axelboc** [comment]
Just had a quick look at all the maps to get an idea of whether this issue is still relevant:

1. The following maps are definitely not up to the standard: `balkan_peninsula`, `ceuta`, `melilla`, `java`, `sumatra`, `saint_helena`, and the five oceans. Two of these are still provided as JPEGs. Note also that some of these notes may be removed from the deck depending on the outcome of #221 and #137.
2. The maps of the continents differ quite significantly from the other maps, but they look decent enough, in my opinion.
3. The locator maps of states and territories are a little inconsistent with one another, but this is really minor. If I recall correctly, they fall mainly into two groups because of a change in the styleguide of the locator maps project on Github. The ones with thick lines for borders are from the older styleguide, I believe. A few maps also follow a subdivision scheme, like the notes of the UK's constituent country.
4. A lot of maps could still use a zoom box, notably those of the Pacific islands. This is the most important point in my opinion (perhaps tied with the first point).
5. Some other maps could use other forms of adjustment. For instance, the map of Kyrgyzstan mentioned in the previous comment could probably be zoomed out a little.

@aplaice et al, do you think we should close this issue and reopen more specific ones for the points above? Do you think all of them are worth fixing? 

I personally feel that this issue has run its course: all the images are properly sourced, the maps of water bodies are fantastic, we've added zoom boxes to all the Antilles maps, ... The map are light years better than they were when I opened this first issue!

**Comment by aplaice** [comment]
> The following maps are definitely not up to the standard: balkan_peninsula, ceuta, melilla, java, sumatra, saint_helena, and the five oceans. Two of these are still provided as JPEGs. Note that some of these notes may be removed from the deck depending on the outcome of #221 and #137.

FWIW (to avoid duplicated effort, just in case somebody becomes inspired) I had been looking into some of these — I have OK-ish replacements for the Balkan Peninsula and Saint Helena, and I've been working on Java and Sumatra (for these two I didn't want to just recolour the Indonesia locator map, as I'm not happy with the central longitude chosen for the projection since it distorts the shape of Indonesia, and in the meantime I discovered a slight bug in my (messy) map rendering script, so I got stalled).

> @aplaice et al, do you think we should close this issue and reopen more specific ones for the points above? Do you think all of them are worth fixing?

I think it makes sense, for a cleaner discussion. I'm not sure if all of them are worth fixing, but it's probably easier to open issues anyway to discuss whether they need to be fixed, there.
